docs/proposals/secure-impersonation.md

```diff
@@ -0,0 +1,636 @@
+Adding another source in the future does not require updating the `flux-source-viewer` object, just creating a new aggregated role.
+
+It's possible to take this one step further and aggregate `flux-source-viewer` into the kubernetes default role for `edit`.
+It would be useful for `view` as well, depending on your security model.
+`flux-apply-editor` and `flux-source-editor` might be a good ClusterRoles to aggregate under `admin` since it's useful for enabling tenants to be
+self-service within their Namespace from their own control-repo.
+
+We need to work out which roles are appropriate to aggregate under any default Kubernetes roles, and whether this should be done with a default
+flux install or taken as an install option.
+
+> Note: `admin` is an aggregated role containing `edit` which contains `view`, but `cluster-admin` is not an aggregated role -- it explicitly allows all resources including any Flux objects.
+
+
+#### kubectl
+`client-go` currently has an issue where `kubectl --as=user --as-group=group` is ignored when using the in-cluster config
+for Pod ServiceAccount tokens.
+This prevents kustomize-controller from using its Pod ServiceAccount to impersonate FluxUsers when applying resources
+by exec'ing `kubectl`.
+The workaround to this bug is to heuristically detect the in-cluster config and supply it explicitly with equivalent
+flags or a kubeConfig file for the token, CA, apiserver-endpoint, and protocol.
+This bug affects all binaries using client-go's flag-options parser, but does not affect libraries like helm that use
+injected clients.
+
+We can maintain a workaround library that shims the kubectl exec or move to using kubectl as a library directly.
+
+
+## Concerns
+#### compatibiliity
+These changes are mostly breaking. Existing Flux deployments are not going to have RBAC configured for every reconciler.
+Some reconciler ServiceAccounts may need `flux-source-viewer` bound to them.
+Others may prefer to update their serviceAccountNames to user.
+Migration/Upgrade documentation will be necessary.
+It's possible to build a tool-assisted migration that modifies resources in a git repo. (`flux migrate <git-repo-path>` ?)
+
+#### performance
+Flux controllers will be using many more Kubernetes client Objects, informers, and caches.
+Storing a map of Flux Users to clients/caches might be a good first step.
+kubeConfig client/caches are not stored currently, which was an explicit decision to support refreshing clients,
+this this doesn't have to be true for all kubeConfigs.
+
+We have done some performance testing in the past, but it would be good to have a strategy/tools to help us
+understand performance regressions.
+
+## Decisions
+Direction should be decided on the following:
+
+- [ ] `flux:user` vs. `system:flux:user` User prefix
+- [ ] `flux:users` vs. `flux:user` Group prefix
+- [ ] default FluxUser name (proposed: "reconciler")
```

Comment:
reconciler or maybe system? Should have a status that writes out the full name.

Reply:
or `flux:user:default`

Reply:
My only concern /w `default` is users conflating it with the default SA in the same NS.
Otherwise it's a pretty good name.

Reply:
Is this supposed to be the default impersonated user the specific controller impersonate when reconciling? Or I am misunderstood something?
Thanks